### PR TITLE
Deprecate import/module name mismatch (ALL CASES)

### PR DESCRIPTION
**Pull request by marler8997**
This PR modifies dmd to verify that the name of a module always matches the name that was used to import it.  Depending on the situation, if the names do not match then this PR will either provide a deprecation message or assert an error.

1. _DEPRECATED_: import a module with a qualified name but the module being imported has no module declaration
```D
// main.d
import foo.bar;
// foo/bar.d
// note the missing module declaration
```
```
 Deprecated: module 'bar' from file foo/bar.d should be imported with 'import bar;'
```

2. _DEPRECATED_: import a module with a qualified name that partially matches the name of the module being imported

```D
// main.d
import foo.bar;
import foo.baz.buz;
// foo/bar.d
module bar;
// foo/baz/buz.d
module baz.buz;
```
```
 Deprecated: module 'bar' from file foo/bar.d should be imported with 'import bar;'
 Deprecated: module 'baz.buz' from file foo/baz/buz.d should be imported with 'import baz.buz;'
```
> NOTE: for this rule to apply, the ends of the imported name and the module name itself must match.

3. _ERROR_: importing a module that matches the filename but does not match the module name
```D
// main.d
import foo;
// foo.d
module bar;
```
```
 Error: module 'bar' from file foo.d must be imported with 'import bar;'
```

4. NOT APPLICABLE/STILL WORKS:  import a module with the correct module name but does not match the filename
```D
// main.d
import bar;
// foo.d
module bar;
```

NOTE: this also addresses a bug found by @timotheecour 

https://github.com/dlang/dmd/pull/7777#discussion_r163939438

NOTE: The DMD test suite itself was using some of this functionality in a fair amount of tests, i.e.
* `test/runnable/imports/link7745b.d`
* `test/fail_compilation/imports/a11919.d`

### CODE REVIEW

File: test/fail_compilation/badimport.d [hunk 1]
```diff
@@ -0,0 +1,4 @@
+// COMPILED_IMPORTS: imports/nomodname.d
+// REQUIRED_ARGS: -Ifail_compilation
+// PERMUTE_ARGS:
+import imports.nomodname;
```
Comment (timotheecour) [inline test/fail_compilation/badimport.d]: no newline ad EOF

Comment (marler8997) [reply test/fail_compilation/badimport.d]: Is that a bad thing?

File: test/fail_compilation/imports/nomodname.d [hunk 2]
```diff
@@ -0,0 +1 @@
+// the point of this module is that it has no module name
```
Comment (timotheecour) [inline test/fail_compilation/imports/nomodname.d]: // intentionally no module declaration

File: src/dmd/root/array.d [hunk 3]
```diff
@@ -313,3 +313,13 @@ unittest
...
 }
+
+bool equals(T)(const(Array!T)* left, const(Array!T)* right)
+{
+    if(left is null || left.dim == 0)
+        return right is null || right.dim == 0;
+    foreach (i; 0 .. left.dim)
+        if ((*left)[i] != (*right)[i])
+	    return false;
+    return true;
```
Comment (timotheecour) [inline src/dmd/root/array.d]: bug? with right.dim > left.dim

File: test/fail_compilation/badimport.d (hunk 1 above)
Comment (timotheecour) [reply test/fail_compilation/badimport.d]: i think it's a convention (lots of editor show an annoying marker when we forget one

File: src/dmd/root/array.d (hunk 3 above)
Comment (marler8997) [reply src/dmd/root/array.d]: shoot, woops

File: src/dmd/dmodule.d [hunk 4]
```diff
@@ -536,6 +537,10 @@ extern (C++) final class Module : Package
         }
         m = m.parse();
 
+        // verify the module name matches the imported module name
+        if (!packages.equals((m.md is null) ? null : m.md.packages))
+            m.error(loc, "from file %s must be imported with 'import %s;'", m.srcfile, m.toPrettyChars());
```
Comment (JinShil) [inline src/dmd/dmodule.d]: Sadly, I think this will introduce a regression, but probably one D needs.

 - This needs to specified in the specification, if it's not already.  
 - This probably needs to go through the deprecation process.  That stinks, but it probably needs to be done.
 - I think this will require a changelog entry

That's just my opinion/perception.  I'd be happy to fast-track this as I think it's the right thing to do to enforce reliable usage of the language, but that's not my call.

**Comment by JinShil** [comment]
Also, I think we require bugzilla entries for all bug fixes.

**Comment by timotheecour** [comment]
@JinShil  found this old bug which might be related: https://issues.dlang.org/show_bug.cgi?id=15086

**Comment by JinShil** [comment]
> found this old bug

Ok, if this PR addresses that bug, prepend "Fix Isssue xxxx" to this PR's title and to the commit message so @dlang-bot will pick it up.

**Comment by jacob-carlborg** [comment]
Is this anything related to declaring a module with a different name than the file? Because that was used as a feature in DMD, before the Objective-C integration was refactored. Previously there were two files, `objc.d` and `objc_stubs.d`, both used the same module declaration: `module ddmd.objc`. This allowed to swap the implementation in the makefiles based on the platform without having to change any imports. See https://github.com/dlang/dmd/commit/9fe35b144f8ad5fbd96489597ebaf979e9bb064a for reference. This is not used anymore and only one `objc.d` file exists.

**Comment by marler8997** [comment]
It's actually about the import name matching the module name.  You can still have a module name that's different from the file name, but this PR makes it so the import name has to match the module name.

**Comment by ibuclaw** [comment]
Yes, I've always considered this to be a feature.

**Comment by marler8997** [comment]
This PR is unrelated to @jacob-carlborg's example (I assume that's what you were referring to @ibuclaw).

File: src/dmd/root/array.d [hunk 5]
```diff
@@ -313,3 +313,15 @@ unittest
     array[2] = 910;
     assert([123, 421, 910, 123, 1, 2, 8, 20, 4, 3] == array.asDArray);
 }
+
```
Comment (UplinkCoder) [inline src/dmd/root/array.d]: why this ?

Comment (marler8997) [reply src/dmd/root/array.d]: Allows you to compare to pointers to array and handles the `null` case for you as well

Comment (marler8997) [reply src/dmd/root/array.d]: ok, changed it to `opEquals`.

**Comment by timotheecour** [comment]
so that module name == import name == _MODULE__ is guaranteed.

On Jan 25, 2018 11:46 PM, "Jonathan Marler" <notifications@github.com>
wrote:

It's actually about the import name matching the module name. You can still
have a module name that's different from the file name, but this PR makes
it so the import name has to match the module name.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/dlang/dmd/pull/7778#issuecomment-360707335>, or mute
the thread

File: src/dmd/dmodule.d [hunk 6]
```diff
@@ -536,6 +537,20 @@ extern (C++) final class Module : Package
...
+        {
+            // in this case there was no module declaration, this means the import
+            // should have no package prefix, but for now we support this with a
+            // depecration message backwards compatibility
+            if (packages && packages.dim > 0)
+                m.deprecation(loc, "from file %s should be imported with 'import %s;'",
+                    m.srcfile.name.toChars(), m.toPrettyChars());
+        }
+        else if (!packages.opEquals(m.md.packages) || ident != m.md.id)
+            m.error(loc, "from file %s must be imported with 'import %s;'",
```
Comment (RazvanN7) [inline src/dmd/dmodule.d]: This should probably go through a deprecation cycle.

File: src/dmd/root/array.d (hunk 5 above)
Comment (wilzbach) [reply src/dmd/root/array.d]: Why not `left.asDArray == right.asDArray`?

File: src/dmd/root/array.d [hunk 7]
```diff
@@ -313,3 +313,15 @@ unittest
...
+bool opEquals(T)(const(Array!T)* left, const(Array!T)* right)
```
Comment (wilzbach) [inline src/dmd/root/array.d]: As always - new symbols:
- should have docstrings
- should use `ref` or `out` instead of pointers (except there's a good reason not to)

File: src/dmd/dmodule.d (hunk 6 above)
Comment (marler8997) [reply src/dmd/dmodule.d]: Maybe.  I think that some variations are more likely to be used out in the wild than others.  Here's my categorization of the various cases:

1. import a module with a qualified name and the module being imported has no module declaration

```D
// main.d
import foo.bar;
// foo/bar.d
// note the missing module declaration
```
This PR doesn't break this case but it does print a deprecation message.

2. import a module with a qualified name that partially matches the name of the module being imported

```D
// main.d
import foo1.bar;
import foo2.bar.baz.
// foo1/bar.d
module bar;
// foo2/baz/buz.d
module baz.buz;
```
Currently this PR makes these cases errors, however, I could see an argument to deprecate these instead.  doing so would be a simple modification to this PR, what do people think?

3. importing a module that matches the filename but does not match the module name
```D
// main.d
import foo;
// foo.d
module bar;
```
This is one case that I don't think we need to deprecate. I say kill it with fire and extreme prejudice.

Comment (marler8997) [reply src/dmd/dmodule.d]: Note that this doesn't break the following case:
```
// main.d
import bar;
// foo.d
module bar;
```

File: src/dmd/root/array.d (hunk 7 above)
Comment (marler8997) [reply src/dmd/root/array.d]: since `left` and `right` could be null, using ref seems odd to me in this case

File: src/dmd/root/array.d (hunk 5 above)
Comment (marler8997) [reply src/dmd/root/array.d]: Would have to be `(*left).asDArray` and `(*right).asDArray`, and this segfaults when left or right is null.

**Comment by timotheecour** [comment]
@jacob-carlborg  @ibuclaw 

> Is this anything related to declaring a module with a different name than the file? Because that was used as a feature in DMD

This is what the flag `-mv=` is for, to allow customizing the file <=> module mapping. This PR doesn't break that.

**Comment by jacob-carlborg** [comment]
> This is what the flag -mv=<package.module>=<filespec> is for, to allow customizing the file <=> module mapping. This PR doesn't break that.

This was done in DMD without that flag.

**Comment by marler8997** [comment]
No activity for 11 days.  I haven't tried to push anyone else to look at this since I'd rather my other PR's receive attention since they need to be merged quickly before the next release, but I have plenty of time if anyone has time to review/approve/request changes.

**Review by wilzbach — APPROVED** [review]
@marler8997 this LGTM. There's still a small nit from @timotheecour, but apart from that I'm just waiting here for @andralex or @WalterBright to have a brief look as seen on the testsuite it might trigger a couple of deprecation messages. However, it's important to realize that (1) it's not breaking any code yet (we never need to follow through with the deprecation and (2) this is in accordance with the D spec.

File: test/compilable/imports/a313templatemixin1.d [hunk 8]
```diff
@@ -1,3 +1,4 @@
+module imports.a313templatemixin1;
```
Comment (andralex) [inline test/compilable/imports/a313templatemixin1.d]: Why is this necessary? We should continue to work if the module declaration is not present.

**Review by andralex — COMMENTED** [review]
It seems this PR forces all modules imported via a path to use a module declaration. Is this correct? If so, it's too much.

File: test/compilable/imports/a313templatemixin1.d (hunk 8 above)
Comment (marler8997) [reply test/compilable/imports/a313templatemixin1.d]: It's one way to go.  You could also add the `imports` directory to the "import path list", i.e.
```
dmd -Icompilable/imports ...
```
then no module declaration is required.

Comment (andralex) [reply test/compilable/imports/a313templatemixin1.d]: @marler8997 can't we just penalize people who enter wrong module information? It seems this PR does that but also adds another restriction, which is more difficult to justify.

**Comment by marler8997** [comment]
> It seems this PR forces all modules imported via a path to use a module declaration. Is this correct? If so, it's too much.

I'm not sure exactly what you mean.  There's a couple different cases.  If you import a module using an "unqualified name", then no module declaration is required.  However, if you import a module with a "qualified name", and that module has no module declaration, then this PR will print a deprecation message. i.e.
```D
// main.d
import foo.bar.baz.buzz;

// foo/bar/baz/buzz.d
// no module declaration (deprecation message)
```

**Comment by andralex** [comment]
@marler8997 if I understand correctly you impose that no module could be imported through two different paths. Isn't that a bit excessive?

**Comment by marler8997** [comment]
> @marler8997 if I understand correctly you impose that no module could be imported through two different paths. Isn't that a bit excessive?

2 different paths is ok, it's 2 different module names that this is preventing.

**Comment by marler8997** [comment]
```D
// main.d
import foo.bar.baz.buz;
// other.d
import baz.buz;
//buz.d

```
Are you saying you think this is OK?

File: test/compilable/imports/a313templatemixin1.d (hunk 8 above)
Comment (marler8997) [reply test/compilable/imports/a313templatemixin1.d]: Sorry, I'm not sure exactly what you're trying to say.  There's 2 cases listed in the original PR message that have been deprecated, and 1 that has been declared to be an error.  Do those use cases categorize the cases you are thinking of?  If so, which ones do you think should not be deprecated/made errors?

Comment (andralex) [reply test/compilable/imports/a313templatemixin1.d]: @marler8997 yah let me discuss exactly on the cases you present in the changelog. Thanks for the neat organization.

File: changelog/deprecate_bad_module_name.dd [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+// foo/bar.d
+module bar;
+
+// foo/baz/buz.d
+module baz.buz;
+---
+The above code will now print:
+
+$(CONSOLE Deprecated: module bar from file foo/bar.d should be imported with 'import bar;'
```
Comment (andralex) [inline changelog/deprecate_bad_module_name.dd]: Alternatively: replace module declaration with `module foo.bar;`

We need to show advantages of this as well. For example show how the current rules lead to a breach in modularity.

My current mental model goes like this: `main.d` uses a dotted import to guide path selection, and that also confers a local name for the imported module. The fact that the module gave itself a non-dotted name is irrelevant to `main.d`. I see nothing wrong in the fact that the module name from the perspective of code inside `foo/bar.d` is different from the module name from the perspective of code inside `main.d`. In fact we have a language construct that messes exactly with that:

```d
import takezat = foo.bar;
```

We'd need a good argument that the above is a bad thing.

File: changelog/deprecate_bad_module_name.dd [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+Deprecated: module baz.buz from file foo/baz/buz.d should be imported with 'import baz.buz;' )
```
Comment (andralex) [inline changelog/deprecate_bad_module_name.dd]: A similar line of reasoning would apply here. We need to build a case on why that's bad.

File: changelog/deprecate_bad_module_name.dd [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+Note that for this rule to apply, the shorter name must completely match the end of the longer name.
+
+Case 3: ERROR: import a module that matches the filename but does not match the module name.
+---
+// main.d
+
```
Comment (andralex) [inline changelog/deprecate_bad_module_name.dd]: Did you mean to add some content to main.d?

File: changelog/deprecate_bad_module_name.dd [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+// foo.d
+module bar;
+---
+The above code will now fail and print:
+
+$(CONSOLE Error: module bar from file foo.d must be imported with 'import bar;')
```
Comment (andralex) [inline changelog/deprecate_bad_module_name.dd]: I don't understand how the file would ever be found. So the compiler is supposed to compile `main.d`. It sees `import bar;`. How does it know it's supposed to look for a file called `foo.d`?

File: changelog/deprecate_bad_module_name.dd [hunk 13]
```diff
@@ -0,0 +1,47 @@
+Deprecate import module name mismatch
+
+Some instances of module name/import name mismatches were deprecated while others are now considered errors.
+
+Case 1: DEPRECATED: import a module name with a qualified name but the module being imported has no module declaration.
```
[the remark this answers is not in the document]
Comment (marler8997) [reply changelog/deprecate_bad_module_name.dd]: > This effectively imposes that modules inside directories (frequent situation) must have a module declaration.

Actually it just means that they must be imported with a "non-qualified" name.

**Comment by andralex** [comment]
>Are you saying you think this should be OK?

@marler8997 what would be the directory structure and compiler flags? Would those be in the same project? (If different projects, I can definitely say that should be okay.)

Overall: this diff renders a use of module names used in a bunch of our own tests as an antipattern. I think this is a serious sign we need to scrutinize the matter appropriately.

File: changelog/deprecate_bad_module_name.dd (hunk 9 above)
Comment (marler8997) [reply changelog/deprecate_bad_module_name.dd]: Interesting.  That's not how I originally understood the module system to work, but your description matches the implementation that I found.

One disadvantage of the current mechanism is that this causes the compiler to load the same module multiple times.  Each unique path used to import the module will result in another instance of the module being instantiated.  I think there's more to say but I'll leave it at that for now.

**Comment by marler8997** [comment]
> Overall: this diff renders a use of module names used in a bunch of our own tests as an antipattern. I think this is a serious sign we need to scrutinize the matter appropriately.

Definitely.  I welcome the scrutiny.  It looks like we'll have to explore this more, maybe I can come up with some use cases that demonstrate the advantages of these restrictions.

File: changelog/deprecate_bad_module_name.dd (hunk 12 above)
Comment (marler8997) [reply changelog/deprecate_bad_module_name.dd]: The file is give on the command line.  Module names not matching file names is a technique already used, though the import statement will match the module name (just no the filename).  You have 3 different names:

1. Filename
2. Import Name
3. Module Name

This PR enforces that "import name" and "module name" should match, but filename does not necessarily need to.  @jacob-carlborg provided an example where this is useful (https://github.com/dlang/dmd/pull/7778#issuecomment-360706239)

**Comment by marler8997** [comment]
> @marler8997 what would be the directory structure and compiler flags? Would those be in the same project? (If different projects, I can definitely say that should be okay.)

I've created the small example project I mentioned and depending on how I compile it/if I add/remove module names, etc. I end up with all kinds of weird errors.  Some cases I get a linker error, some cases I get a compilation error, and some cases I don't get any error, but each case is very non-intuitive and requires intimiate knowledge of how the compiler works to understand what's going on.

I argue that the ability to import a module using different names severely breaks the module system.  Normally if people play nice and don't do weird things then it's ok, but this PR is preventing those weird abuses which in my opinion have to utility and only serve to add confusion to the module system.

**Comment by marler8997** [comment]
Let's look at one of the weird things that can happen with the current state of the "unverified" module system. Take the following 2 files:
```D
// main.d
import foo.bar;
void main() { barfunc(); }

// foo/bar.d
void barfunc() { }
```

If you compile "foo/bar.d" seperately like this:
```
dmd -c foo/bar.d
```

Then you can successfully compile the full program via
```
dmd main.d bar.o
```

However, if you compile it fully from source you will get a compilation error:
```
dmd main.d foo/bar.d
```
```
Error: module bar from foo/bar.d must be imported with `import bar;`
```

This error occurs becase the module `foo/bar.d` is being loaded twice with 2 different names.  The first time it is loaded because it is given on the command line, and since it has no module declaration, the compiler gives it the name `bar`.  But then the module gets imported with the name `foo.bar`, however, the compiler cannot change the name of the module and all existing reference to it so an error must be asserted.  In the first case where the module is precompiled, the module only gets loaded once from the import so it will be given the proper name of `foo.bar` and the compilation succeeds.

This difference in behavior couples the module system with how the files are compiled.  Compiling some combinations of files will succeed while others will not.

This PR breaks this coupling by forbidding this behavior altogether.  No matter what combination you compile the files in, you will always get the following error:
```
Error: module bar from foo/bar.d must be imported with `import bar;` or given a name with `module foo.bar;`
```

This example demonstrates the core problem with supporting the first 2 cases that this PR aims to deprecate. However, note that this example is just one manifestation of the core problem.  The same core problem could also cause different files to be imported from the same import statement, which can cause nonsensical compilation/linker errors which are very difficult to root cause.

**Comment by marler8997** [comment]
P.S. Many of the tests in the DMD test suite will only compile if you compile the imports separately from the test file itself.  If you compile them all together in one invocation, you will get an import error.  This demonstrates a real example where not verifying import/module names puts a requirement on what order/how you compile your modules together.

**Comment by WalterBright** [comment]
The fact that this PR requires 80-90 files in the test suite to be edited casts serious doubt on the viability of this change w.r.t. breaking existing code.

**Comment by marler8997** [comment]
Editing all those files is not required. I'll modify the PR to remove those changes to demonstrate.

**Comment by wilzbach** [comment]
Better submit a new PR, s.t. you don't loose these changes ;-)

**Comment by marler8997** [comment]
Since it looks like it's going to take some work to reach consensus on cases 1 and 2, I'm going to split this PR.  I'm going to push a new PR that only addresses case 3 (which addresses issue 15086).  Once that is merged I can create a new PR that addresses case 1 and 2.  This limits the scope of each PR making them easier to discuss and prevents conflation of the individual points for each case.

**Comment by marler8997** [comment]
Closing this one in favor of (https://github.com/dlang/dmd/pull/7878) and potentially a new PR in the future for cases 1 and 2.
